Add resources overview document. (#345)
nitrictech/docs: docs/resources-overview.md

Nitric provides cloud native building blocks that make it simple to declare the resources you need as part of your application code.

> The Nitric deployment engine will run through your code at deploy time and interpret resources that are declared as part of your application and create them in the cloud you are pushing to.

Reviewer: ```suggestion
The Nitric deployment engine will run through your code at deploy time, interpreting resources that are declared as part of your application and creating them in the cloud you are pushing to.
```

## Rules

> There are a few rules to keep in mind when declaring nitric resources as part of your application

Reviewer: ```suggestion
There are a few rules to keep in mind when declaring Nitric resources as part of your application.
```

### Don't declare resources in runtime code

> Nitric needs to be aware of resources at deploy time so they can be deployed appropriately. Declaring resources at runtime mean the resource won't be declared when deploying your application and therefore it will not be provisioned to the cloud.

Reviewer: ```suggestion
Nitric needs to be aware of resources at deploy time so they can be deployed appropriately. Declaring resources at runtime means the resource won't be declared when deploying your application. Consequently, the resource will not be provisioned to the cloud.
```

Reviewer: Avoid run on sentence.

> The Nitric deployment engine does not evaluate runtime code at deploy time as this could result in unintended behaviour/side-effects.

Reviewer: ```suggestion
The Nitric deployment engine does not evaluate runtime code at deploy time as this could result in unintended behaviours or side-effects.
```

A working example:

```typescript
import { api, bucket } from '@nitric/sdk';

// ✅ This declaration will work
const files = bucket('files').for('reading');

const publicApi = api('public');

> publicApi.get('/files/:name', () => {

Reviewer: ```suggestion
publicApi.get('/files/:name', (ctx) => {
```

>   // ❌ This declaration will not work, as this resource will not exist

Reviewer: Reference the runtime reasoning again.

```suggestion
  // ❌ This declaration will not work, as this is only called at runtime.
```

### ❌ Avoid declaring permissions for shared resources

Creating resource permissions in the same context as the resources can make those permissions leak. This is demonstrated in the below example:

```typescript
// lib/resources.ts
import { api, topic, bucket } from '@nitric/sdk';

export const publicApi = api('public');

export const bucketOne = bucket('bucket-one').for('reading');
export const bucketTwo = bucket('bucket-two').for('reading');
```

```typescript
// functions/function-one.ts
import { publicApi, bucketOne } from '../lib/resources';

publicApi.get('bucket-one/file/:name', (ctx) => {
  // do something with the file
});
```

```typescript
// functions/function-two.ts
import { publicApi, bucketTwo } from '../lib/resources';

publicApi.get('bucket-two/file/:name', (ctx) => {
  // do something with the file
});
```

In this scenario both `function-one` and `function-two` have read access to both buckets, even though each function is only using one of the buckets. This is because they are declared in the same context and evaluated at the same time in each of the functions.

> Resource permissions should always be declared for in the scope of a single function, unless the permissions required by one or more functions are identical.

Reviewer: ```suggestion
Resource permissions should always be declared in the scope of a single function unless the permissions required by one or more functions are identical.
```

> > This practice can break principal of least priveledge in infrastructure deployments

Reviewer: ```suggestion
> This practice can break the principal of least privilege in infrastructure deployments
```
